client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,308 @@
+/// filtered out.
+pub struct BlockExecutor<Block: BlockT, Client> {
+	client: Arc<Client>,
+	block: Block::Hash,
+	targets: Option<String>,
+	storage_keys: Option<String>,
+}
+
+impl<Block, Client> BlockExecutor<Block, Client>
+	where
+		Block: BlockT + 'static,
+		Client: HeaderBackend<Block> + BlockBackend<Block> + ProvideRuntimeApi<Block>
+		+ Send + Sync + 'static,
+		Client::Api: Metadata<Block>,
+{
+	/// Create a new `BlockExecutor`
+	pub fn new(
+		client: Arc<Client>,
+		block: Block::Hash,
+		targets: Option<String>,
+		storage_keys: Option<String>,
+	) -> Self {
+		Self { client, block, targets, storage_keys }
+	}
+
+	/// Execute block, recording all spans and events belonging to `Self::targets`
+	/// and filter out events which do not have the keys starting with one of the
+	/// prefixes in `Self::storage_keys`.
+	pub fn trace_block(&self) -> Result<TraceBlockResponse, String> {
+		tracing::debug!(target: "state_tracing", "Tracing block: {}", self.block);
+		// Prepare the block
+		let id = BlockId::Hash(self.block);
+		let mut header = self.client.header(id)
+			.map_err(|e| format!("Invalid block Id: {:?}", e))?
+			.ok_or("Header not found".to_string())?;
+		let extrinsics = self.client.block_body(&id)
+			.map_err(|e| format!("Invalid block Id: {:?}", e))?
+			.ok_or("Extrinsics not found".to_string())?;
+		tracing::debug!(target: "state_tracing", "Found {} extrinsics", extrinsics.len());
+		let parent_hash = header.parent_hash().clone();
+		let parent_id = BlockId::Hash(parent_hash.clone());
+		// Pop digest else RuntimePanic due to: 'Number of digest items must match that calculated.'
+		header.digest_mut().pop();
+		let block = Block::new(header, extrinsics);
+
+		let targets = if let Some(t) = &self.targets { t } else { DEFAULT_TARGETS };
+		let storage_keys = if let Some(s) = &self.storage_keys {
+			s
+		} else {
+			DEFAULT_STORAGE_KEYS
+		};
+		let spans = Mutex::new(HashMap::new());
+		let events = Mutex::new(Vec::new());
+		let block_subscriber = BlockSubscriber::new(targets, spans, events);
+		let dispatch = Dispatch::new(block_subscriber);
+
+		{
+			let dispatcher_span = tracing::debug_span!(target: "state_tracing",
+				"execute_block", extrinsics_len = block.extrinsics().len());
+			let _guard = dispatcher_span.enter();
+			if let Err(e) = dispatcher::with_default(&dispatch, || {
+				let span = tracing::info_span!(
+					target: TRACE_TARGET,
+					"trace_block",
+				);
+				let _enter = span.enter();
+				self.client.runtime_api().execute_block(&parent_id, block)
+			}) {
+				return Err(format!("Error executing block: {:?}", e));
+			}
+		}
+
+		let block_subscriber = dispatch.downcast_ref::<BlockSubscriber>()
+			.ok_or("Cannot downcast Dispatch to BlockSubscriber after tracing block")?;
+		let mut span_data: Vec<SpanDatum> = block_subscriber.spans
+			.lock()
+			.drain()
+			.map(|(_, s)| SpanDatum::from(s))
+			.collect();
+		span_data.sort_by(|a, b| a.start_time.cmp(&b.start_time));
```

Comment:
Is there a way to do this sort without collecting twice? My only idea was to add start_time into Span. (Moving comment out of code.)